### ISSUE TITLE
Investigate: -6 pts MMLU divergence on 26B-A4B 4-bit MoE vs Python mlx-lm

### ISSUE THREAD

**Issue by VincentGourbin** [VincentGourbin/gemma-4-swift-mlx#27]
## Summary

Lors de la validation du portage sur 5 modèles Gemma 4 ([BENCHMARKS.md](BENCHMARKS.md)), une régression isolée a été identifiée sur le path quantisé MoE du modèle 26B-A4B :

| Config | Swift MMLU plain | Python MMLU plain | Δ | Swift MMLU Pro | Python MMLU Pro | Δ |
|---|---|---|---|---|---|---|
| 26B-A4B-4bit | 57.0% | 63.0% | **-6.0** | 44.8% | 47.6% | -2.8 |
| 26B-A4B-bf16 | 59.0% | 57.0% | +2.0 | 51.4% | 48.6% | +2.8 |

**Le bug est isolé au path quantisé** : sur bf16, Swift égale ou bat Python (cohérent avec validation des autres modèles non-MoE : 31B-4bit exact match, 12B-bf16 dans le bruit, etc.). Sur 4-bit, Swift perd 6 pts.

## Diagnostic effectué

Le code Swift `Gemma4Router.swift` et `Gemma4Experts.swift` (utilisant `mlx-swift-lm/SwitchLayers.swift`) est quasi-identique au Python `mlx-lm/gemma4_text.py` et `switch_layers.py`. Les deux appellent la **même fonction C++ `gather_qmm`** dans la lib MLX.

```swift
// Swift mlx-swift-lm/SwitchLayers.swift L181-192
MLX.gatherQuantizedMM(x, self.weight, scales: self.scales, biases: self.biases,
    rhsIndices: indices, transpose: true, groupSize: self.groupSize,
    bits: self.bits, mode: mode, sortedIndices: sortedIndices)
```

```python
# Python mlx-lm/switch_layers.py
mx.gather_qmm(x, self["weight"], self["scales"], self.get("biases"),
    rhs_indices=indices, transpose=True, group_size=self.group_size,
    bits=self.bits, mode=self.mode, sorted_indices=sorted_indices)
```

Tentative de port verbatim de la version Python Router (softmax-sur-top-k + rms_norm fusionné) a **dégradé** l'accuracy de -3 pts au lieu d'améliorer → cause non triviale.

## Hypothèses (par ordre de probabilité)

1. **Précision intermédiaire des scales/biases** : Python `mx.quantize` retourne scales en `float16` par défaut. Swift peut les charger/caster différemment. Sur 24 layers × 8 experts × 3 projections = **576 quantized matmuls par token**, l'accumulation de bruit est non-triviale.

2. **Tri `_gather_sort` / `gatherSort`** : ordre des indices après tri peut différer subtilement entre les deux orchestrations.

3. **Binding Swift ↔ MLX C++** : marshalling des arrays peut introduire des casts bf16↔float32 différents de Python.

4. **Lazy eval MLX** : ordre de matérialisation différent peut bouger l'argmax sur cas borderline (effet observé sur logits proches).

## Plan d'investigation proposé

1. **Instrumenter une seule couche** : dump outputs `gate_proj`, `up_proj`, `down_proj` quantisés en Swift et Python sur le même input (token et top-k indices identiques).
2. **Comparer numériquement les sorties** : trouver à quelle profondeur le drift commence à dépasser le seuil de changement d'argmax.
3. **Si drift dans `gather_qmm` lui-même** → reporter à [`ml-explore/mlx`](https://github.com/ml-explore/mlx) (C++).
4. **Si drift dans l'orchestration** (sort/scatter, cast) → reporter à [`ml-explore/mlx-swift-lm`](https://github.com/ml-explore/mlx-swift-lm).

## Reproduction

```bash
# Bench 4-bit (-6 pts gap)
./.build/xcode/Build/Products/Release/gemma4-cli eval-mmlu \
  --model-path ~/Library/Caches/models/mlx-community/gemma-4-26b-a4b-it-4bit \
  --dataset /tmp/mmlu_5shot.json
# Swift: 57/100 = 57.0%
# Python (cd /tmp/benchwork && python3 mmlu_python.py): 63/100 = 63.0%

# Bench bf16 (validation pas de bug archi)
./.build/xcode/Build/Products/Release/gemma4-cli eval-mmlu \
  --model-path ~/Library/Caches/models/mlx-community/gemma-4-26b-a4b-it-bf16 \
  --dataset /tmp/mmlu_5shot.json
# Swift: 59/100 = 59.0%
# Python: 57/100 = 57.0%
```

## Impact

- **Mineur** : affecte uniquement le modèle 26B-A4B en quantisation 4-bit (le mlx-community 4-bit officiel).
- **Pas critique** : le portage est validé sur 4 autres modèles + sur 26B-A4B bf16.
- **À ne pas confondre** : la régression -25 pts MMLU sur 12B en OTF 4-bit pure documentée ailleurs est un effet inhérent au modèle (pas un bug de port).

## Pointeurs code

- `Sources/Gemma4Swift/TextModel/Gemma4Router.swift` — Router Swift (validé bf16)
- `Sources/Gemma4Swift/TextModel/Gemma4Experts.swift` — Experts wrapper (validé bf16)
- `mlx-swift-lm/Libraries/MLXLMCommon/SwitchLayers.swift` L150-200 — QuantizedSwitchLinear suspect
- `mlx-lm/mlx_lm/models/switch_layers.py` — référence Python (équivalente)
